### PR TITLE
Custom Asset Bridging after OZ fixes were applied

### PR DESCRIPTION
**Pull request by Raid5594**
# What ❔

## Checklist

- [ ] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.

### CODE REVIEW

File: l2-contracts/contracts/common/libraries/DataEncoding.sol [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    /// @param _assetData The asset data that has to be encoded.
+    /// @return The encoded asset data.
+    function encodeNTVAssetId(bytes32 _assetData) internal view returns (bytes32) {
+        return keccak256(abi.encode(block.chainid, NATIVE_TOKEN_VAULT_VIRTUAL_ADDRESS, _assetData));
+    }
+
+    /// @notice Encodes the asset data by combining chain id, NTV as asset deployment tracker and asset data.
+    /// @param _tokenAaddress The address of token that has to be encoded (asset data is the address itself).
+    /// @return The encoded asset data.
+    function encodeNTVAssetId(address _tokenAaddress) internal view returns (bytes32) {
```
Comment (kelemeno) [inline l2-contracts/contracts/common/libraries/DataEncoding.sol:83]: typo

File: l1-contracts/contracts/bridge/L1NativeTokenVault.sol [hunk 2]
```diff
@@ -154,91 +174,29 @@ contract L1NativeTokenVault is IL1NativeTokenVault, IL1AssetHandler, Ownable2Ste
...
         // solhint-disable-next-line func-named-parameters
-        _bridgeMintData = abi.encode(amount, _prevMsgSender, _l2Receiver, getERC20Getters(l1Token), l1Token);
-        // solhint-disable-next-line func-named-parameters
+        _bridgeMintData = DataEncoding.encodeBridgeMintData(
+            amount,
+            _prevMsgSender,
+            _l2Receiver,
+            getERC20Getters(l1Token),
+            l1Token
+        ); // solhint-disable-next-line func-named-parameters
```
Comment (StanislavBreadless) [inline l1-contracts/contracts/bridge/L1NativeTokenVault.sol:183]: do we need to disable it here? maybe we could make the event accept named params

File: l1-contracts/contracts/bridge/L1NativeTokenVault.sol [hunk 3]
```diff
@@ -154,91 +174,29 @@ contract L1NativeTokenVault is IL1NativeTokenVault, IL1AssetHandler, Ownable2Ste
         chainBalance[_chainId][l1Token] += amount;
 
         // solhint-disable-next-line func-named-parameters
```
Comment (StanislavBreadless) [inline l1-contracts/contracts/bridge/L1NativeTokenVault.sol:176]: same goes here, maybe use named params?

File: l1-contracts/contracts/bridge/L1SharedBridge.sol [hunk 4]
```diff
@@ -219,16 +224,36 @@ contract L1SharedBridge is IL1SharedBridge, ReentrancyGuard, Ownable2StepUpgrade
...
     }
 
-    /// @notice Sets the asset handler address for a given asset ID.
-    /// @dev No access control on the caller, as msg.sender is encoded in the assetId.
-    /// @param _assetData In most cases this parameter is bytes32 encoded token address. However, it can include extra information used by custom asset handlers.
-    /// @param _assetHandlerAddress The address of the asset handler, which will hold the token of interest.
-    function setAssetHandlerAddressInitial(bytes32 _assetData, address _assetHandlerAddress) external {
-        address sender = msg.sender == address(nativeTokenVault) ? NATIVE_TOKEN_VAULT_VIRTUAL_ADDRESS : msg.sender;
-        bytes32 assetId = keccak256(abi.encode(uint256(block.chainid), sender, _assetData));
+    /// @notice Used to Sets the assed deployment tracker address for given asset data.
```
Comment (StanislavBreadless) [inline l1-contracts/contracts/bridge/L1SharedBridge.sol:227]:
```suggestion
    /// @notice Used to set the assed deployment tracker address for given asset data.
```

File: l1-contracts/contracts/bridge/L1SharedBridge.sol [hunk 5]
```diff
@@ -239,37 +264,37 @@ contract L1SharedBridge is IL1SharedBridge, ReentrancyGuard, Ownable2StepUpgrade
...
-            (_assetId, _assetAddressOnCounterPart)
+            (_assetId, _assetHandlerAddressOnCounterPart)
         );
 
         L2TransactionRequestDirect memory request = L2TransactionRequestDirect({
             chainId: _chainId,
             l2Contract: l2BridgeAddress[_chainId],
-            mintValue: _mintValue, // l2 gas + l2 msg.value the bridgehub will withdraw the mintValue from the base token bridge for gas
-            l2Value: 0, // L2 msg.value, this contract doesn't support base token deposits or wrapping functionality, for direct deposits use bridgehub
+            mintValue: _mintValue, // l2 gas + l2 msg.Value the bridgehub will withdraw the mintValue from the base token bridge for gas
```
Comment (StanislavBreadless) [inline l1-contracts/contracts/bridge/L1SharedBridge.sol:289]: I am not sure using big `V` was intended here

```suggestion
            mintValue: _mintValue, // l2 gas + l2 msg.value the bridgehub will withdraw the mintValue from the base token bridge for gas
```

File: l1-contracts/contracts/bridge/L1SharedBridge.sol [hunk 6]
```diff
@@ -390,34 +353,31 @@ contract L1SharedBridge is IL1SharedBridge, ReentrancyGuard, Ownable2StepUpgrade
 
         if (encodingVersion == 0x01) {
             (assetId, transferData) = abi.decode(_data[1:], (bytes32, bytes));
+            require(
+                assetHandlerAddress[assetId] != address(nativeTokenVault),
+                "ShB: new encoding format not yet supported for NTV"
+            );
         } else {
             (assetId, transferData) = _handleLegacyData(_data, _prevMsgSender);
```
Comment (StanislavBreadless) [inline l1-contracts/contracts/bridge/L1SharedBridge.sol:361]: Please add a comment on why `encodingVersion == 0x01` can not accidentally collide with legacy encoding

File: l1-contracts/contracts/bridge/L1SharedBridge.sol [hunk 7]
```diff
@@ -559,82 +487,59 @@ contract L1SharedBridge is IL1SharedBridge, ReentrancyGuard, Ownable2StepUpgrade
...
-            address l1Token = nativeTokenVault.tokenAddress(_assetId);
-            (uint256 amount, address prevMsgSender) = abi.decode(_transferData, (uint256, address));
-            bytes32 txDataHash = keccak256(abi.encode(prevMsgSender, l1Token, amount));
-            require(dataHash == txDataHash, "ShB: d.it not hap");
+            // Determine if the given dataHash matches the calculated legacy transaction hash.
+            bool isLegacyTxDataHash = _isLegacyTxDataHash(_depositSender, _assetId, _assetData, dataHash);
+            // If the dataHash matches the legacy transaction hash, skip the next step.
+            // Otherwise, perform the check using the new transaction data hash encoding.
+            if (!isLegacyTxDataHash) {
+                bytes32 txDataHash = this.encodeTxDataHash(false, _depositSender, _assetId, _assetData);
```
Comment (StanislavBreadless) [inline l1-contracts/contracts/bridge/L1SharedBridge.sol:495]: nit: we could create an internal method for situations like this. This is a free way to save up on gas to not do additional calls/abi encoding/decoding etc.

File: l2-contracts/contracts/bridge/L2NativeTokenVault.sol [hunk 8]
```diff
@@ -95,53 +97,60 @@ contract L2NativeTokenVault is IL2NativeTokenVault, Ownable2StepUpgradeable {
...
-    function bridgeMint(uint256 _chainId, bytes32 _assetId, bytes calldata _transferData) external payable override {
+    /// @param _data The abi.encoded transfer data.
+    function bridgeMint(uint256 _chainId, bytes32 _assetId, bytes calldata _data) external payable override onlyBridge {
         address token = tokenAddress[_assetId];
-        (address _l1Sender, uint256 _amount, address _l2Receiver, bytes memory erc20Data, address originToken) = abi
-            .decode(_transferData, (address, uint256, address, bytes, address));
-        address expectedToken = l2TokenAddress(originToken);
+        (
+            uint256 _amount,
+            address _l1Sender,
```
Comment (StanislavBreadless) [inline l2-contracts/contracts/bridge/L2NativeTokenVault.sol:108]: is it a breaking change? Before the data used to start with l1Sender, but now it starts with amount

File: l2-contracts/contracts/bridge/L2SharedBridge.sol [hunk 9]
```diff
@@ -142,42 +151,35 @@ contract L2SharedBridge is IL2SharedBridge, ILegacyL2SharedBridge, Initializable
...
+        bytes calldata _data
     ) external override {
-        // onlyBridge {
-        bytes32 assetId = keccak256(abi.encode(L1_CHAIN_ID, NATIVE_TOKEN_VAULT_VIRTUAL_ADDRESS, _l1Token));
+        bytes32 assetId = keccak256(
+            abi.encode(L1_CHAIN_ID, NATIVE_TOKEN_VAULT_VIRTUAL_ADDRESS, bytes32(uint256(uint160(_l1Token))))
+        );
         // solhint-disable-next-line func-named-parameters
-        bytes memory data = abi.encode(_l1Sender, _amount, _l2Receiver, erc20Data, _l1Token);
+        bytes memory data = DataEncoding.encodeBridgeMintData(_amount, _l1Sender, _l2Receiver, _data, _l1Token);
```
Comment (StanislavBreadless) [inline l2-contracts/contracts/bridge/L2SharedBridge.sol:182]: same about this one, looks like a breaking change

File: l2-contracts/contracts/bridge/interfaces/IL2SharedBridge.sol [hunk 10]
```diff
@@ -22,15 +23,15 @@ interface IL2SharedBridge {
...
-    function finalizeDeposit(bytes32 _assetId, bytes calldata _data) external;
+    function finalizeDeposit(bytes32 _assetId, bytes calldata _transferData) external;
 
-    function withdraw(bytes32 _assetId, bytes calldata _data) external;
+    function withdraw(bytes32 _assetId, bytes calldata _transferData) external;
 
     function l1Bridge() external view returns (address);
 
-    function assetHandlerAddress(bytes32 _assetId) external view returns (address);
+    function assetHandlerAddress(bytes32 assetId) external view returns (address);
```
Comment (kelemeno) [inline l2-contracts/contracts/bridge/interfaces/IL2SharedBridge.sol:32]: missing underscore, _assetId .

File: l2-contracts/contracts/bridge/interfaces/IL2NativeTokenVault.sol [hunk 11]
```diff
@@ -21,11 +22,11 @@ interface IL2NativeTokenVault is IL2AssetHandler {
         uint256 amount
     );
 
-    event L2TokenBeaconUpdated(address l2TokenBeacon, bytes32 l2TokenProxyBytecodeHash);
+    event L2TokenBeaconUpdated(address indexed l2TokenBeacon, bytes32 indexed l2TokenProxyBytecodeHash);
 
     function l2Bridge() external view returns (IL2SharedBridge);
 
-    function tokenAddress(bytes32 _assetId) external view returns (address);
+    function tokenAddress(bytes32 assetId) external view returns (address);
```
Comment (kelemeno) [inline l2-contracts/contracts/bridge/interfaces/IL2NativeTokenVault.sol:29]: Is the underscore missing on purpose? _assetId

File: l2-contracts/contracts/bridge/L2SharedBridge.sol [hunk 12]
```diff
@@ -142,42 +151,35 @@ contract L2SharedBridge is IL2SharedBridge, ILegacyL2SharedBridge, Initializable
...
         address _l2Receiver,
         address _l1Token,
         uint256 _amount,
-        bytes calldata erc20Data
+        bytes calldata _data
     ) external override {
-        // onlyBridge {
-        bytes32 assetId = keccak256(abi.encode(L1_CHAIN_ID, NATIVE_TOKEN_VAULT_VIRTUAL_ADDRESS, _l1Token));
+        bytes32 assetId = keccak256(
+            abi.encode(L1_CHAIN_ID, NATIVE_TOKEN_VAULT_VIRTUAL_ADDRESS, bytes32(uint256(uint160(_l1Token))))
```
Comment (kelemeno) [inline l2-contracts/contracts/bridge/L2SharedBridge.sol:179]: Is the bytes32(uint256(uint160( ))) wrapping required or not? It is sometimes there, sometimes not

File: l2-contracts/contracts/bridge/interfaces/IL2NativeTokenVault.sol (hunk 11 above)
Comment (Raid5594) [reply l2-contracts/contracts/bridge/interfaces/IL2NativeTokenVault.sol:29]: also a mapping

File: l2-contracts/contracts/bridge/L2NativeTokenVault.sol [hunk 13]
```diff
@@ -113,9 +113,7 @@ contract L2NativeTokenVault is IL2NativeTokenVault, Ownable2StepUpgradeable {
 
         if (token == address(0)) {
             address expectedToken = _calculateCreate2TokenAddress(originToken);
-            bytes32 expectedAssetId = keccak256(
-                abi.encode(_chainId, NATIVE_TOKEN_VAULT_VIRTUAL_ADDRESS, bytes32(uint256(uint160(originToken))))
-            );
+            bytes32 expectedAssetId = keccak256(abi.encode(_chainId, NATIVE_TOKEN_VAULT_VIRTUAL_ADDRESS, originToken));
```
Comment (StanislavBreadless) [inline l2-contracts/contracts/bridge/L2NativeTokenVault.sol:116]: BTW, we already have the `DataEncoding` library for such cases.

File: l2-contracts/contracts/bridge/L2SharedBridge.sol (hunk 12 above)
Comment (kelemeno) [reply l2-contracts/contracts/bridge/L2SharedBridge.sol]: Maybe this should be moved to the DataEncoding as well?

Comment (Raid5594) [reply l2-contracts/contracts/bridge/L2SharedBridge.sol]: removed it everywhere

File: l2-contracts/contracts/common/libraries/DataEncoding.sol [hunk 14]
```diff
@@ -67,20 +67,20 @@ library DataEncoding {
     /// @param _sender The asset deployment tracker address.
     /// @return The encoded asset data.
     function encodeAssetId(address _tokenAaddress, address _sender) internal view returns (bytes32) {
-        return keccak256(abi.encode(block.chainid, _sender, _tokenAaddress));
+        return keccak256(abi.encode(L1_CHAIN_ID, _sender, _tokenAaddress));
```
Comment (StanislavBreadless) [inline l2-contracts/contracts/common/libraries/DataEncoding.sol:70]: the same `DataEncoding` library has two methods named `encodeAssetId` with different params and very different logic, where in one of those  `L1_CHAIN_ID` is used and in the other `block.chainid` is used. They should have semantically same logic

Secondly, it seems like we hardcode  `L1_CHAIN_ID = 1` here which wont work on local network/testnets

File: l2-contracts/contracts/bridge/L2NativeTokenVault.sol [hunk 15]
```diff
@@ -31,7 +35,7 @@ contract L2NativeTokenVault is IL2NativeTokenVault, Ownable2StepUpgradeable {
     /// @dev Bytecode hash of the proxy for tokens deployed by the bridge.
     bytes32 internal l2TokenProxyBytecodeHash;
 
-    mapping(bytes32 assetId => address tokenAddress) public override tokenAddress;
+    mapping(bytes32 _assetId => address tokenAddress) public override tokenAddress;
```
Comment (StanislavBreadless) [inline l2-contracts/contracts/bridge/L2NativeTokenVault.sol:38]: please remove underscore

File: l2-contracts/contracts/bridge/L2NativeTokenVault.sol [hunk 16]
```diff
@@ -119,10 +119,10 @@ contract L2NativeTokenVault is IL2NativeTokenVault, Ownable2StepUpgradeable {
         if (token == address(0)) {
             address expectedToken = _calculateCreate2TokenAddress(originToken);
             bytes32 expectedAssetId = DataEncoding.encodeNTVAssetId(L1_CHAIN_ID, originToken);
-            if (_assetId != expectedAssetId) {
-                // Make sure that a NativeTokenVault sent the message
-                revert AssetIdMismatch(expectedAssetId, _assetId);
-            }
+            // if (_assetId != expectedAssetId) {
```
Comment (StanislavBreadless) [inline l2-contracts/contracts/bridge/L2NativeTokenVault.sol:122]: ?

**Review by StanislavBreadless — APPROVED** [review]

**Pull request merged**